scripts-dev/schema_versions.py

```diff
@@ -0,0 +1,143 @@
+
+                        # Bail early.
+                        break
+        else:
+            # SCHEMA_COMPAT_VERSION is sometimes across multiple lines, the easist
+            # thing to do is exec the code. Luckily it has only ever existed in
+            # a file which imports nothing else from Synapse.
+            locals: Dict[str, Any] = {}
+            exec(schema_file.data_stream.read().decode("utf-8"), {}, locals)
+            schema_version = locals["SCHEMA_VERSION"]
+            schema_compat_version = locals.get("SCHEMA_COMPAT_VERSION")
+
+    return schema_version, schema_compat_version
+
+
+def get_tags(repo: git.Repo) -> Iterator[git.Tag]:
+    """Return an iterator of tags sorted by version."""
+    tags = []
+    for tag in repo.tags:
+        # All "real" Synapse tags are of the form vX.Y.Z.
+        if not tag.name.startswith("v"):
+            continue
+
+        # There's a weird tag from the initial react UI.
+        if tag.name == "v0.1":
+            continue
+
+        try:
+            tag_version = version.parse(tag.name)
+        except version.InvalidVersion:
+            # Skip invalid versions.
+            continue
+
+        # Skip pre- and post-release versions.
+        if tag_version.is_prerelease or tag_version.is_postrelease or tag_version.local:
+            continue
+
+        tags.append((tag_version, tag))
+
+    # Sort based on the version number (not lexically).
+    return (tag for _, tag in sorted(tags, key=lambda t: t[0]))
+
+
+def calculate_version_chart() -> str:
+    repo = git.Repo(path=".")
+
+    schema_version = None
+    schema_compat_version = None
+
+    # Map of schema version -> Synapse versions at that schema.
+    schema_versions = defaultdict(list)
+
+    # Find ranges of versions which are compatible with a schema version.
+    #
+    # There are two modes of operation:
+    #
+    # 1. If schema_compat_version is None, then Synapse can only move to a new
+    #    version with schema_version >= its current version.
+    # 2. If schema_compat_version is *not* None, then Synapse can move to a new
+    #    version with schema version >= schema_compat_version.
+    #
+    # See https://github.com/matrix-org/synapse/pull/9933 which was included in v1.37.0.
+    for tag in get_tags(repo):
+        schema_version, schema_compat_version = get_schema_versions(tag)
+
+        # If a schema compat version is given, prefer that over the schema version.
+        schema_versions[schema_compat_version or schema_version].append(tag.name)
+
+    # Generate a table of which maps a version to the version it can be rolled back to.
+    result = "| Synapse version | Backwards compatible version |\n"
+    result += "|-----------------|------------------------------|\n"
+    for synapse_versions in schema_versions.values():
+        result += f"| {synapse_versions[-1]: ^15} | {synapse_versions[0]: ^28} |\n"
+
+    return result
+
+
+if __name__ == "__main__":
```

Comment:
You can no longer easily run this to splat it into your terminal, which is annoying. I should probably make it do something more reasonable if there's no input?

Reply:
Maybe if sys.argv is empty / only has the program name, assume it's a manual invocation rather than from the book?